sw/host/sphincsplus/lib.rs

```diff
@@ -2,16 +2,149 @@
+    let mut pk = [0u8; SPX_PUBLIC_KEY_BYTES];
+    let mut sk = [0u8; SPX_SECRET_KEY_BYTES];
+    let err_code =
+        // SAFETY: the buffers here are all fixed-length arrays of the size expected by the C code.
+        unsafe { crypto_sign_seed_keypair(pk.as_mut_ptr(), sk.as_mut_ptr(), seed.0.as_ptr()) };
+    anyhow::ensure!(
+        err_code == 0,
+        "SPHINCS+ seeded key pair generation failed with error code {err_code}."
+    );
+    Ok((SpxPublicKey(pk), SpxSecretKey(sk)))
+}
+
+// Generate a new random keypair.
+pub fn spx_keypair_generate() -> anyhow::Result<(SpxPublicKey, SpxSecretKey)> {
+    let mut pk = [0u8; SPX_PUBLIC_KEY_BYTES];
+    let mut sk = [0u8; SPX_SECRET_KEY_BYTES];
+    let err_code =
+        // SAFETY: the buffers here are all fixed-length arrays of the size expected by the C code.
+        unsafe { crypto_sign_keypair(pk.as_mut_ptr(), sk.as_mut_ptr()) };
+    anyhow::ensure!(
+        err_code == 0,
+        "SPHINCS+ random key pair generation failed with error code {err_code}."
+    );
+    Ok((SpxPublicKey(pk), SpxSecretKey(sk)))
+}
+
+// Generate a detached signature for the message using the secret key.
+pub fn spx_sign(sk: &SpxSecretKey, msg: &Vec<u8>) -> anyhow::Result<SpxSignature> {
+    let mut sig = [0u8; SPX_SIGNATURE_BYTES];
+    let mut sig_bytes_written = 0;
+    let err_code =
+        // SAFETY: the signature and secret key buffers here are fixed-length arrays of the size
+        // expected by the C code, and the message buffer is passed along with its length. The
+        // signature is always the same length, but the implementation returns the number of bytes
+        // written as part of the result; we check this value later in the non-error case against
+        // the expected length.
+        unsafe {
+        crypto_sign_signature(
+            sig.as_mut_ptr(),
+            &mut sig_bytes_written,
+            msg.as_ptr(),
+            msg.len(),
+            sk.0.as_ptr(),
+        )
+    };
+    anyhow::ensure!(
+        err_code == 0,
+        "SPHINCS+ signature generation failed with error code {err_code}."
+    );
+    anyhow::ensure!(
+        sig_bytes_written == sig.len(),
+        "SPHINCS+ signature has unexpected length: {sig_bytes_written} != {}.",
+        sig.len()
+    );
+    Ok(SpxSignature(sig))
+}
+
+// Verify a detached signature and return true if the signature is valid.
+pub fn spx_verify(pk: &SpxPublicKey, sig: &SpxSignature, msg: &Vec<u8>) -> bool {
+    let err_code =
+        // SAFETY: the signature and public key buffers here are fixed-length arrays of the size
+        // expected by the C code, and the message buffer is passed along with its length.
+        unsafe {
+        crypto_sign_verify(
+            sig.0.as_ptr(),
+            sig.0.len(),
+            msg.as_ptr(),
+            msg.len(),
+            pk.0.as_ptr(),
+        )
+    };
+    err_code == 0
```

Comment:
You may want to look at the verification functions in some of the RustCrypto libraries we use.  IIRC, it's fairly typical to return `Result<(), Error>` from those functions, with `Ok(())` indicating verification success.

Reply:
You're right, should be addressed now.